diff --git a/android/build.gradle b/android/build.gradle
index 20f41eaae06df2959ca8a4c8235af3ceb6c07bc2..3b6e7f2ecf5e8a9b5c3dd85837cceb7afaba6919 100644
--- a/android/build.gradle
+++ b/android/build.gradle
@@ -58,6 +58,13 @@ def resolveReactNativeDirectory() {
 
 if (isNewArchitectureEnabled()) {
     apply plugin: 'com.facebook.react'
+
+    def reactNativeDirectory = resolveReactNativeDirectory()
+
+    react {
+        reactNativeDir = file(reactNativeDirectory)
+        codegenDir = file("${file(reactNativeDirectory).getParentFile()}/react-native-codegen")
+    }
 }
 apply plugin: 'com.android.library'
 apply plugin: 'kotlin-android'
